Composer/packages/client/src/pages/botProject/createPublishTarget.tsx

```diff
@@ -0,0 +1,307 @@
+        />
+      </form>
+    );
+  }, [name, targetType, targetTypes, errorMessage]);
+
+  const PageEditProfile = useMemo(() => {
+    return (
+      <div style={{ width: '60%' }}>
+        {FormInPage}
+        <Fragment>
+          {selectedType?.instructions && <p>{selectedType.instructions}</p>}
+          <div css={label}>{formatMessage('Publish Configuration')}</div>
+          <JsonEditor
+            key={targetType}
+            editorSettings={userSettings.codeEditor}
+            height={200}
+            schema={selectedType?.schema}
+            value={config}
+            onChange={setConfig}
+          />
+        </Fragment>
+      </div>
+    );
+  }, [targetType, selectedType, userSettings, FormInPage, config]);
+
+  const PageContent = useMemo(() => {
+    switch (page) {
+      case PageTypes.AddProfile:
+        return (
+          <Fragment>
+            <div style={{ width: '60%', minHeight: '300px' }}>{FormInPage}</div>
+            <Separator css={separator} />
+            <DialogFooter>
+              <DefaultButton text={formatMessage('Cancel')} onClick={props.closeDialog} />
+              <PrimaryButton
+                disabled={nextDisabled}
+                text={formatMessage('Next')}
+                onClick={async () => {
+                  if (isShowAuthDialog(true)) {
+                    setShowAuthDialog(true);
+                  } else {
+                    setPage(PageTypes.ConfigProvision);
+                  }
+                }}
+              />
+            </DialogFooter>
+          </Fragment>
+        );
+      case PageTypes.EditProfile:
+        return (
+          <Fragment>
+            {PageEditProfile}
+            <Separator css={separator} />
+            <DialogFooter>
+              <DefaultButton text={formatMessage('Cancel')} onClick={props.closeDialog} />
+              <PrimaryButton disabled={saveDisabled} text={formatMessage('Save')} onClick={submit} />
+            </DialogFooter>
+          </Fragment>
+        );
+      case PageTypes.ConfigProvision:
+        if (selectedType?.bundleId) {
+          // render custom plugin view
+          let publishSurfaceStyles;
+          switch (selectedType.extensionId) {
+            case 'pva-publish-composer':
+              publishSurfaceStyles = pvaPublishSurface;
+              break;
+            case 'azurePublish':
+              publishSurfaceStyles = azurePublishSurface;
+              break;
+            default:
+              publishSurfaceStyles = defaultPublishSurface;
+              break;
+          }
+          return (
+            <Fragment>
+              <div css={publishSurfaceStyles}>
+                <PluginHost
+                  bundleId={selectedType.bundleId}
+                  pluginName={selectedType.extensionId}
+                  pluginType="publish"
+                />
+              </div>
+            </Fragment>
+          );
+        }
+    }
+    return null;
+  }, [
+    currentUser,
```

Comment:
Almost all states are added to `useMemo`'s deps, I'm curious about how will it affect the performance? Why not define more React FC rather than warpping them with `useMemo`?

Reply:
Yeah, Can do it later, but I think the performance is the same as warpping in Function Component.